### ISSUE TITLE
Re-enable disabled compiler warnings and upgrade boost. Ref: NCC-2016-007

### ISSUE THREAD

**Issue by nathan-at-least** [zcash/zcash#944]
Now that we've made a pragmatic improvement with #939, I'd like to start follow up improvements using [the technique @daira suggested](https://github.com/zcash/zcash/issues/911#issuecomment-217049144), quoted here to save you a click:

> Could we use
> ```
> #pragma GCC diagnostic push
> #pragma GCC diagnostic warning "-Wno-deprecated-declarations"
> #include <boost/smart_ptr/shared_ptr.hpp>
> #include <boost/smart_ptr/scoped_ptr.hpp>
> #pragma GCC diagnostic pop
> ```
> in a header that gets included first?

For this ticket: figure out which disabled warning class has the fewest warning-sources in boost, then re-enable that warning, then use the pragma technique to make the result compile.

Then, we can either continue to the next warning category, or stop at any point and our static-analysis assurances will still be better off.

Note: For this process we'll leave ``-Werror`` enabled, and incrementally re-enable more and more static analysis checks.

**Comment by defuse** [comment]
The boost warnings seem isolated to one specific header file in boost (although that header file may be included by many other of boost's header files). So one way to attack this would be to patch boost's header files themselves with those `#pragma`s instead of putting them around our `#include` statements.

**Comment by daira** [comment]
I like that idea.

**Comment by daira** [comment]
Boost used this approach themselves in the fix to https://svn.boost.org/trac/boost/ticket/11622 , which was committed 11 months ago. We should just upgrade, and then remove our own use of ``-Wno-deprecated-declarations``.

**Comment by nathan-at-least** [comment]
Upgrading boost is problematic. In the past we had to *downgrade* for C++11 compatibility. Caveat emptor.

**Comment by daira** [comment]
Recent boost works fine with C++11, and won't even compile with C++03. We can't stay stuck in the past; that's a recipe for security flaws.

**Comment by daira** [comment]
Boost 1.62.0 is [in beta](http://www.boost.org/users/history/version_1_62_0.html) so it *might* be released in time for us to use it.

**Comment by str4d** [comment]
Boost 1.62.0 is scheduled for release [September 28](http://www.boost.org/development/index.html), so we will be able to use it. I will leverage the latest 1.62.0 beta release in my PR for #867 so we can test against it, and then I will upgrade to 1.62.0 next week after it is released, before merging the PR for our beta 2.

**Comment by nathan-at-least** [comment]
Note: let's ensure that 1.62.0 fixes #1241 (which requires having the right security analysis tool in place).

**Comment by str4d** [comment]
@rcseacord @daira other than upgrading to latest Boost, what else exactly was required to close this issue? I can add it to #1429.

**Comment by str4d** [comment]
- ec3bd70b6d6ef8be901a5523a401e333c52e7bbe added `-Wno-deprecated-declarations` (before we had `-Werror`.
- be1bae14e8a769966b833026ed47a43f5250a4b9 hid new Boost warnings on GCC 6 using `-Wno-placement-new -Wno-terminate`.

#1429 now upgrades Boost to 1.62, but I am unable to remove `-Wno-deprecated-declarations` because we still get deprecation warnings about `auto_ptr`. The patch linked in the ticket @daira referenced in https://github.com/zcash/zcash/issues/944#issuecomment-239417799 is definitely merged in Boost 1.62, but it only fixed the Boost shared pointers library. The warning I am running into is:

```
In file included from /home/str4d/dev/zcash/zcash/depends/x86_64-unknown-linux-gnu/include/boost/spirit/home/classic/core/non_terminal/grammar.hpp:21:0,
                 from /home/str4d/dev/zcash/zcash/depends/x86_64-unknown-linux-gnu/include/boost/spirit/home/classic/core.hpp:42,
                 from /home/str4d/dev/zcash/zcash/depends/x86_64-unknown-linux-gnu/include/boost/spirit/include/classic_core.hpp:11,
                 from json/json_spirit_reader_template.h:19,
                 from asyncrpcoperation.h:20,
                 from asyncrpcqueue.h:8,
                 from asyncrpcqueue.cpp:5:
/home/str4d/dev/zcash/zcash/depends/x86_64-unknown-linux-gnu/include/boost/spirit/home/classic/core/non_terminal/impl/grammar.ipp: In member function ‘boost::spirit::classic::impl::grammar_helper<GrammarT, DerivedT, ScannerT>::definition_t& boost::spirit::classic::impl::grammar_helper<GrammarT, DerivedT, ScannerT>::define(const grammar_t*)’:
/home/str4d/dev/zcash/zcash/depends/x86_64-unknown-linux-gnu/include/boost/spirit/home/classic/core/non_terminal/impl/grammar.ipp:159:18: error: ‘template<class> class std::auto_ptr’ is deprecated [-Werror=deprecated-declarations]
             std::auto_ptr<definition_t>
                  ^
In file included from /usr/include/c++/5/memory:81:0,
                 from asyncrpcoperation.h:13,
                 from asyncrpcqueue.h:8,
                 from asyncrpcqueue.cpp:5:
/usr/include/c++/5/bits/unique_ptr.h:49:28: note: declared here
   template<typename> class auto_ptr;
```

This shows up several times but references the same file. I think the fix for this is to apply a patch that includes the header file Boost added with the `#pragma` (although that adds a dependency from the Boost spirit library to the shared pointers library that may not exist).

**Comment by daira** [comment]
That's unfortunate. Yes, let's patch out that #pragma.

**Comment by str4d** [comment]
I have added a patch for Boost to fix the `auto_ptr` warnings in #1429, and re-enabled all disabled compiler warnings. Someone needs to confirm that there are no warnings on GCC 6.

**Comment by str4d** [comment]
I installed the following GCC 6 version:

```
g++-6 (Ubuntu 6.2.0-3ubuntu11~16.04) 6.2.0 20160901
```

I did a `make clean`, deleted `depends/x86_64-unknown-linux-gnu/` and `depends/work/build/`, and then applied the following patch:

```
diff --git a/depends/builders/default.mk b/depends/builders/default.mk
index f097db6..2edbe37 100644
--- a/depends/builders/default.mk
+++ b/depends/builders/default.mk
@@ -1,5 +1,5 @@
-default_build_CC = gcc
-default_build_CXX = g++
+default_build_CC = gcc-6
+default_build_CXX = g++-6
 default_build_AR = ar
 default_build_RANLIB = ranlib
 default_build_STRIP = strip
diff --git a/depends/hosts/default.mk b/depends/hosts/default.mk
index 6f60d6b..68c19af 100644
--- a/depends/hosts/default.mk
+++ b/depends/hosts/default.mk
@@ -1,5 +1,5 @@
-default_host_CC = $(host_toolchain)gcc
-default_host_CXX = $(host_toolchain)g++
+default_host_CC = $(host_toolchain)gcc-6
+default_host_CXX = $(host_toolchain)g++-6
 default_host_AR = $(host_toolchain)ar
 default_host_RANLIB = $(host_toolchain)ranlib
 default_host_STRIP = $(host_toolchain)strip
diff --git a/depends/hosts/linux.mk b/depends/hosts/linux.mk
index 31748d6..91d14ac 100644
--- a/depends/hosts/linux.mk
+++ b/depends/hosts/linux.mk
@@ -10,15 +10,15 @@ linux_debug_CXXFLAGS=$(linux_debug_CFLAGS)
 linux_debug_CPPFLAGS=-D_GLIBCXX_DEBUG -D_GLIBCXX_DEBUG_PEDANTIC
 
 ifeq (86,$(findstring 86,$(build_arch)))
-i686_linux_CC=gcc -m32
-i686_linux_CXX=g++ -m32
+i686_linux_CC=gcc-6 -m32
+i686_linux_CXX=g++-6 -m32
 i686_linux_AR=ar
 i686_linux_RANLIB=ranlib
 i686_linux_NM=nm
 i686_linux_STRIP=strip
 
-x86_64_linux_CC=gcc -m64
-x86_64_linux_CXX=g++ -m64
+x86_64_linux_CC=gcc-6 -m64
+x86_64_linux_CXX=g++-6 -m64
 x86_64_linux_AR=ar
 x86_64_linux_RANLIB=ranlib
 x86_64_linux_NM=nm
```

After this, `./zcutils/build.sh` built #1429 successfully with all three flags above removed (and `-Werror` still enabled). So I'm declaring that #1429 definitely fixes this issue.
